src/plugins/certauth/test/main.c

```diff
@@ -0,0 +1,239 @@
+ * COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT,
+ * INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED
+ * OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <k5-int.h>
+#include "krb5/certauth_plugin.h"
+
+struct krb5_certauth_moddata_st {
+    int initialized;
+};
+
+/* Test module 1 returns OK with an indicator. */
+static krb5_error_code
+test1_authorize(krb5_context context, krb5_certauth_moddata moddata,
+                krb5_data cert, krb5_principal princ, void *opts,
+                void *db_entry, char ***authinds_out)
+{
+    krb5_error_code ret;
+    char **ais = NULL;
+
+    *authinds_out = NULL;
+    ais = k5calloc(2, sizeof(*ais), &ret);
+    if (ais == NULL)
+        return ret;
+
+    ais[0] = strdup("test1");
+    if (ais[0] == NULL) {
+        free(ais);
+        return ENOMEM;
+    }
+
+    *authinds_out = ais;
+    return 0;
+}
+
+static void
+test_free_ind(krb5_context context, krb5_certauth_moddata moddata,
+                                    char **authinds)
+{
+    size_t i;
+    if (authinds == NULL)
+        return;
+    for (i = 0; authinds[i] != NULL; i++) {
+        free(authinds[i]);
+    }
+    free(authinds);
+}
+
+/* A basic moddata test. */
+static krb5_error_code
+test2_init(krb5_context context, krb5_certauth_moddata *moddata_out)
+{
+    krb5_certauth_moddata mod;
+    mod = calloc(1, sizeof(*mod));
+    if (mod == NULL)
+        return ENOMEM;
+    mod->initialized = 1;
+    *moddata_out = mod;
+    return 0;
+}
+
+static void
+test2_fini(krb5_context context, krb5_certauth_moddata moddata)
+{
+   free(moddata);
+}
+
+static krb5_error_code
+has_cn(krb5_context context, krb5_data cert, const char *name)
+{
+    krb5_error_code ret;
+    krb5_boolean match = FALSE;
+    uint8_t cntag[6] = "\x06\x03\x55\x04\x03\x0c";
+    char *c = NULL;
+    struct k5buf buf;
+    size_t name_len, b_len, c_left;
+    char b_start;
+
+    /* Construct a DER search string of CN tag + length tag + name to
+     * check for princ as CN. */
+    k5_buf_init_dynamic(&buf);
+    k5_buf_add_len(&buf, cntag, sizeof(cntag));
+    name_len = strlen(name);
+    k5_buf_add_len(&buf, &name_len, 1);
+    k5_buf_add_len(&buf, name, name_len);
+    ret = k5_buf_status(&buf);
+    if (ret)
+        goto out;
+
+    b_start = *(char *)buf.data;
+    b_len = buf.len;
+    c_left = cert.length;
+
+    /* Check for the CN needle in the certificate haystack. */
+    c = memchr(cert.data, b_start, c_left);
+    while (c != NULL) {
+        c_left = cert.length - ((c + 1) - cert.data);
+        if (b_len > c_left)
```

Comment:
Does this work if the needle is present at the last possible position in the haystack?